### ISSUE TITLE
How to achieve Offline XSUAA Token (JWT) Validation for SpringBoot Application on Cloud Foundry?

### ISSUE THREAD

**Issue by saad-r** [SAP/cloud-security-xsuaa-integration#649]
**Tech Stack**
- client library: java-security-test (2.10.5) & xsuaa-spring-boot-starter (2.10.5)
- Environment: Cloud Foundry 
- app-router with XSUAA binding
- spring-boot application with spring-security

**What works** 
The authentication flow works perfectly in production environment
**What doesn't work**
 But what is the best way to achieve Offline Token Validation for XSUAA JWT.

**More Context**
I am aware of the examples provided in [java-security-test](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/java-security-test) but they seem to make sense for mocking Unit Test Cases, would you recommend the same for Local Development and still how would you achieve it ?
It would be of great help if you can share a working code snippet as well.

**Comment by liga-oz** [comment]
Hi @saad-r,

In your local environment what is missing is the Xsuaa service configuration, that is provided in VCAP_SERVICES environment variable when deployed on cloud. You can set these values manually in your application.yml file. See example [here](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/spring-security#overwrite-identity-service-properties) for `spring-security` library.

Kind REgards,
Liga

**Comment by saad-r** [comment]
Thank for responding @liga-oz .
I don't have a challenge in generating the token locally with the below code snippet

public String getToken() {

        
        Token token = JwtGenerator.getInstance(Service.XSUAA, "client-id")                
                .withClaimValue(TokenClaims.XSUAA.GRANT_TYPE,"dev@mail.com")              
                .withClaimValue("email", "dev@mail.com")
                .withClaimValue("user_name", "dev@mail.com")
                .withClaimValue("given_name", "first name")
                .withClaimValue("family_name", "last name")
                .withScopes("READ:OBJECT","WRITE:OBJECT")               
                .createToken();

        String authorizationHeaderValue = "Bearer " + token.getTokenValue();

        return authorizationHeaderValue;

    }
But how can I validate it, the below SecurityConfiguration would only validate if it can access XSUAA service which is not the case when you generate a mock token locally.

@Configuration
@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    @Autowired
    XsuaaServiceConfiguration xsuaaServiceConfiguration;

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.sessionManagement().sessionCreationPolicy(SessionCreationPolicy.STATELESS).and().authorizeRequests()
                .antMatchers("/api/v1").hasAuthority("READ:OBJECT")
                .antMatchers("/localtoken").permitAll()
                .antMatchers("/*").authenticated().anyRequest().denyAll()
                .and()
                .oauth2ResourceServer()
                .jwt()
                .jwtAuthenticationConverter(getJwtAuthoritiesConverter());
    }

    Converter<Jwt, AbstractAuthenticationToken> getJwtAuthoritiesConverter() {
        TokenAuthenticationConverter converter = new TokenAuthenticationConverter(xsuaaServiceConfiguration);
        converter.setLocalScopeAsAuthorities(true); // not applicable in case of multiple xsuaa bindings!

        return converter;
    }
}

**Comment by liga-oz** [comment]
Hi @saad-r ,

If I understood you correctly, you generate your token and with this token want to call the app running locally?

To achieve that, one option I can think of is to provide your own server that would respond to JWK calls or to provide your own implementation of `OAuth2TokenKeyService` class that is used in your custom `JwtDecoder` bean that you configure with your custom `CombiningValidator`. However, the latest approach using  `JwtValidatorBuilder.withOAuth2TokenKeyService` is deprecated.

```java
@Bean
public JwtDecoder jwtDecoder(XsuaaServiceConfiguration xsuaaConfig,
				IdentityServiceConfiguration identityConfig) {
	CombiningValidator<Token> xsuaaValidator = JwtValidatorBuilder.getInstance(xsuaaConfig)
			.withOAuth2TokenKeyService(yourOAuth2TokenKeyServiceImpl)
			.build();
	return new HybridJwtDecoder(xsuaaValidator, null)
}
```

May I ask for what purpose you want to use this testing setup?

**Comment by saad-r** [comment]
Firstly thank you for your patience.

Maybe let me rephrase this.

There are 2 cases
1. **Testing** : This is document well enough with examples in the java-security-test example. I don't have a problem with this. 
2. **Making calls to App via POSTMAN** :  If I am running the app locally on my system which you would generally do by for e.g.  `mvn spring-boot:run` how can I make calls via Postman to the app's local server that would run say on `localhost:8080/someendpoint/` as then Postman would require a header with a Bearer Token.

**Comment by liga-oz** [comment]
For the 2. case I would suggest to better use a real Xsuaa service that you consume from your local machine instead of creating custom `JwtDecoder` bean as it most likely won't depict productive use anyway.

**Comment by saad-r** [comment]
Okay then this might be the way to do this.
Can you please share any code snippet / more guidance on this.

**Comment by liga-oz** [comment]
You don't have to run it locally, you can access it in cloud over the wide area network. Please check the [hybrid sample](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/samples/spring-security-hybrid-usage#deployment-to-cloud-foundry) it demonstrates how to create xsuaa service and obtain a token from it. When your app runs locally then you need to define VCAP_SERVICES configuration by yourself that will reflect the created Xsuaa service configuration:

> Hi @saad-r,
> 
> In your local environment what is missing is the Xsuaa service configuration, that is provided in VCAP_SERVICES environment variable when deployed on cloud. You can set these values manually in your application.yml file. See example [here](https://github.com/SAP/cloud-security-xsuaa-integration/tree/main/spring-security#overwrite-identity-service-properties) for `spring-security` library. Then of course you would need to mock responses or configure a mockserver.
> 
> Kind Regards,
> Liga

**Comment by saad-r** [comment]
Okay got thanks a ton.

**Comment by saurabhtalreja** [comment]
Hello @saad-r 
I'm facing the same issue, can you please share a sample source code where you've implemented it for local postman use case and as well as for production on CF?
